analyzers/src/SonarAnalyzer.CSharp/Rules/UseModelBinding.cs

```diff
@@ -0,0 +1,253 @@
+    {
+        context.RegisterCompilationStartAction(compilationStartContext =>
+        {
+            var (argumentDescriptors, propertyAccessDescriptors) = GetDescriptors(compilationStartContext.Compilation);
+            if (argumentDescriptors.Any() || propertyAccessDescriptors.Any())
+            {
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    // If the user overrides any action filters, model binding may not be working as expected. Then we do not want to raise on expressions that originate from parameters.
+                    // See the OverridesController.Undecidable test cases for details.
+                    var hasOverrides = false;
+                    var controllerCandidates = new ConcurrentStack<ReportCandidate>(); // In SymbolEnd, we filter the candidates based on the overriding we learn on the go.
+                    if (symbolStartContext.Symbol is INamedTypeSymbol namedType && namedType.IsControllerType())
+                    {
+                        symbolStartContext.RegisterCodeBlockStartAction<SyntaxKind>(codeBlockStart =>
+                            hasOverrides |= RegisterCodeBlockActions(codeBlockStart, argumentDescriptors, propertyAccessDescriptors, controllerCandidates));
+                    }
+                    symbolStartContext.RegisterSymbolEndAction(symbolEnd =>
+                    {
+                        foreach (var candidate in controllerCandidates.Where(x => !(hasOverrides && x.OriginatesFromParameter)))
+                        {
+                            symbolEnd.ReportIssue(Diagnostic.Create(Rule, candidate.Location, candidate.Message));
+                        }
+                    });
+                }, SymbolKind.NamedType);
+            }
+        });
+    }
+
+    private bool RegisterCodeBlockActions(SonarCodeBlockStartAnalysisContext<SyntaxKind> codeBlockStart,
+        ArgumentDescriptor[] argumentDescriptors, MemberDescriptor[] propertyAccessDescriptors,
+        ConcurrentStack<ReportCandidate> controllerCandidates)
+    {
+        if (codeBlockStart.OwningSymbol is IMethodSymbol method && IsOverridingFilterMethods(method))
+        {
+            // We do not want to raise in ActionFilter overrides. The SymbolEndAction needs to be made aware, that there are
+            // ActionFilter overrides, so it can filter out some candidates.
+            return true;
+        }
+        // Within a single code block, access via constant and variable keys could be mixed
+        // We only want to raise, if all access were done via constants
+        var allConstantAccess = true;
+        var codeBlockCandidates = new ConcurrentStack<ReportCandidate>();
+        if (argumentDescriptors.Any())
+        {
+            codeBlockStart.RegisterNodeAction(nodeContext =>
+            {
+                var argument = (ArgumentSyntax)nodeContext.Node;
```

Comment:
It is three lines of code more and does the same thing (We do not care for the additional allocation: We use SymbolStart to limit ourselves to controllers already). It may help with readability. If you prefer it strongly I will change it but for the moment I keep it.